proposal/package-importer.d.ts.md

```diff
@@ -0,0 +1,484 @@
+  > guarantees that `resolved` will have one of these extensions.
+
+* Return `text`, `syntax`, and `resolved`.
+
+[parsing a URL]: https://url.spec.whatwg.org/#concept-url-parser
+[resolving a `pkg:` URL]: #node-algorithm-for-resolving-a-pkg-url
+
+
+## Procedures
+
+### Node Algorithm for Resolving a `pkg:` URL
+
+This algorithm takes a URL with scheme `pkg:` named `url`, and an optional URL
+`previousURL`. It returns a canonical file path or null.
+
+* Let `fullPath` be `url`'s path.
+
+* Let `packageName` be the result of [resolving a package name] with `fullPath`,
+  and `subPath` be the path without the `packageName`.
+
+* Let `packageRoot` be the result of [resolving the root directory for a
+  package] with `packageName` and `previousURL`.
+
+* If a `package.json` file does not exist at `packageRoot`, throw an error.
+
+* Let `packageManifest` be the result of parsing the `package.json` file at
+  `packageRoot` as [JSON].
+
+* Let `resolved` be the result of [resolving package exports] with `sass` as the
+  condition, `packageRoot`, `subpath`, and `packageManifest`.
+
+  * If `resolved` has the scheme `file:` and an extension of `sass`, `scss` or
+    `css`, return it.
+
+  * Otherwise, if `resolved` is not null, throw an error.
+
+* Let `resolved` be the result of [resolving package exports] with `style` as the
+  condition, `packageRoot`, `subpath`, and `packageManifest`.
+
+  * If `resolved` has the scheme `file:` and an extension of `sass`, `scss` or
+    `css`, return it.
+
+  * Otherwise, if `resolved` is not null, throw an error.
+
+* If `subPath` is empty, return the result of [resolving package root values].
+
+* Let `resolved` be `subPath` resolved relative to `packageRoot`.
+
+* Return the result of [resolving a `file:` URL] with `resolved`.
+
+
+[previous URL]: ../accepted/prev-url.d.ts.md
+[Resolving package exports]: #resolving-package-exports
+[resolving package root values]: #resolving-package-root-values
+[resolving a package name]: #resolving-a-package-name
+[JSON]: https://datatracker.ietf.org/doc/html/rfc8259
+[parsing a URL]: https://url.spec.whatwg.org/#concept-url-parser
+[resolving the root directory for a package]: #resolving-the-root-directory-for-a-package
+[resolving a `file:` URL]: ../spec/modules.md#resolving-a-file-url
+
+### Resolving a package name
+
+This algorithm takes a string, `path`, and returns the portion that identifies
+the Node package.
+
+* If `path` starts with `@`, it is a scoped package. Return the first 2 [URL path
+  segments], including the separating `/`.
+
+* Otherwise, return the first URL path segment.
+
+### Resolving the root directory for a package
+
+This algorithm takes a string, `packageName`, and an optional absolute URL
+`previousURL`, and returns an absolute URL to the root directory for the most
+proximate installed `packageName`.
+
+* Let `parentURL` be the value of `previousURL` if it is defined, otherwise the
+  URL to the current directory.
```

Comment:
I agree, this seemed like a magic value appearing. I like your suggestion, and added this in d2eb4f9. I'd welcome your feedback on if I need to be more specific on using `require.main.filename`.

Reply:
I do think this needs to be more specific. The algorithm itself shouldn't specify what its argument will be; the point of an argument is that it's set by the callers that _invoke_ the algorithm. You also can't just add a new algorithm to invoking an importer, because "an importer" is already a well-defined interface [specified here](https://github.com/sass/sass/blob/main/spec/modules.md#importer).

Instead, you need an additional layer that _creates_ the Node resolution importer given an entrypoint URL. (Think of it like a function that returns a closure in JavaScript.) Then in the Node resolution algorithm itself, you can invoke resolving a `pkg:` URL as Node "with `baseURL` set to the canonical URL of the [current source file](https://github.com/sass/sass/blob/main/spec/spec.md#current-source-file) that contained the load if it's a `file:` URL, and `entrypointURL` otherwise". (I'm stealing the verbiage from [the prev-url spec](https://github.com/sass/sass/blob/main/accepted/prev-url.d.ts.md).)

You'll also want to be more specific in the section that defines `pkgImporter`. It's still written like user-facing documentation, but it needs to specify how to construct the Node resolution importer (specifically by passing in `require.main.filename`) _and_ where it goes in the list of importers (probably after user-defined importers but before load paths).

Reply:
I updated this in 0c7ef34. I followed the pattern from loadPaths generating File Importers, adding the concept of an "associated" value. Instead of adding a separate instantiation procedure, I have the `pkgImporter` procedure creating a Node Package Importer with an associated `entryPointURL`. Let me know if this works.

I also added the ordering in the global imports list as part of the generic Package Imports requirements.